proxy/src/control/discovery.rs

```diff
@@ -499,7 +526,110 @@ where T: HttpService<RequestBody = BoxBody, ResponseBody = RecvBody>,
     }
 }
 
-// ===== impl RxError =====
+// ===== impl DstLabels ====
+
+impl DstLabels {
+
+    #[inline]
+    fn new<I, S>(labels: I) -> Option<Self>
+    where
+        I: IntoIterator<Item=(S, S)>,
+        S: fmt::Display,
+    {
+        let mut labels = labels.into_iter();
+
+        if let Some((k, v)) = labels.next() {
+            // format the first label pair without a leading comma, since we
+            // don't know where it is in the output labels at this point.
+            let mut s = format!("dst_{}=\"{}\"", k, v);
+
+            // format subsequent label pairs with leading commas, since
+            // we know that we already formatted the first label pair.
+            for (k, v) in labels {
+                write!(s, ",dst_{}=\"{}\"", k, v)
+                    .expect("writing to string should not fail");
+            }
+
+            Some(DstLabels(Arc::from(s)))
+        } else {
+            // the iterator is empty; return None.
+            None
+        }
+
+    }
+
+}
+
+impl fmt::Display for DstLabels {
+
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+
+}
+
+impl Labeled<SocketAddr> {
+    fn from_pb(pb: WeightedAddr, set_labels: &Arc<HashMap<String, String>>)
+               -> Option<Self> {
+        let inner = pb.addr.and_then(pb_to_sock_addr)?;
+        let label_iter =
+            set_labels.as_ref()
+                .iter()
+                .chain(pb.metric_labels.iter());
+        let metric_labels = DstLabels::new(label_iter);
+        Some(Labeled { metric_labels, inner, })
+    }
+}
+
+
+impl<T> Labeled<T> {
+    pub fn none(inner: T) -> Self {
+        Self { metric_labels: None, inner }
+    }
+
+    fn label<U>(&self, inner: U) -> Labeled<U> {
+        Labeled {
+            metric_labels: self.metric_labels.as_ref().cloned(),
+            inner,
+        }
+    }
+}
+
+impl<T> ops::Deref for Labeled<T> {
+    type Target = T;
+    fn deref(&self) -> &T {
+        &self.inner
+    }
+}
+
+impl<T> Borrow<T> for Labeled<T> {
```

Comment:
Unless it's painfully obvious to all other readers, it might be worth documenting why this is implemented: that it enables Labeled types to be stored in maps/sets without factoring labels into hash computation.

Reply:
Will do. I think it's a fairly well-known idiom but not so much so that it shouldn't be commented on.